CefGlue.Common.Shared/Serialization/JsonConverters/ObjectJsonConverter.cs

```diff
@@ -1,60 +1,170 @@
+                    case JsonTokenType.EndArray:
+                        state.Pop();
+                        continue;
 
-                case JsonTokenType.StartArray:
-                    return JsonDocument.ParseValue(ref reader).Deserialize<object[]>(options);
+                    case JsonTokenType.StartObject:
+                        value = ReadComplexObject(ref reader, state, referencesMap);
+                        break;
 
-                case JsonTokenType.StartObject:
-                    return JsonDocument.ParseValue(ref reader).Deserialize<Dictionary<string, object>>(options);
+                    case JsonTokenType.EndObject:
+                        state.Pop();
+                        continue;
+                }
 
-                default:
-                    return JsonDocument.ParseValue(ref reader).RootElement.Clone();
+                SetObjectPropertyOrArrayIndex(currentState, value);
+            } while (reader.Read());
+
+            if (state.Count > 1 && state.Pop().ObjectHolder != root)
+            {
+                throw new InvalidOperationException("Invalid json format - missing enclosing EndArray or EndObject token(s).");
             }
+
+            return root.First();
+        }
+
+        private static void SetObjectPropertyOrArrayIndex(ReadState state, object value)
+        {
+            var obj = state.ObjectHolder;
+            if (state.PropertyName != null)
+            {
+                ((Dictionary<string, object>)obj)[state.PropertyName] = value;
+            }
+            else if (state.ArrayIndex != null)
+            {
+                ((object[])obj)[state.ArrayIndex.Value] = value;
+                state.ArrayIndex++;
+            }
+        }
+
+        private object ReadComplexObject(ref Utf8JsonReader reader, Stack<ReadState> state, IDictionary<string, object> referencesMap)
+        {
+            var tempReader = reader;
+            object obj = null;
+            ReadState newState = null;
+
+            tempReader.ReadToken(JsonTokenType.StartObject);
+
+            if (tempReader.TokenType == JsonTokenType.EndObject)
+            {
+                // empty object
+                obj = new object();
+            }
+            else
+            {
+                // peek first prop name
+                var propName = tempReader.ReadPropertyName();
+                switch (propName)
+                {
+                    case JsonAttributes.Ref:
+                        reader.ReadToken(JsonTokenType.StartObject);
+                        reader.ReadPropertyName();  // skip the $ref
+                        var refId = reader.GetString();
+                        obj = referencesMap[refId];
+                        break;
+
+                    case JsonAttributes.Id:
+                        reader.ReadToken(JsonTokenType.StartObject);
+                        reader.ReadPropertyName(); // skip the $id
+                        var id = tempReader.ReadString();
```

Comment:
i guess its the reader here, at leats its not consistent with the branch above
```suggestion
                        var id = reader.ReadString();
```

Reply:
It can't be the reader, because we're peeking for the $values property - if it's not, we can't consume that reader position, because it will be consumed in the while cycle continuation.

Reply:
but something is not right, the branch above doesn't do the same... and they should be consistent because the scenario is very similar

Reply:
Not exactly the same as the branch above, because that branch uses the `reader.GetString()`, which doesn't consume the position. And if I were to use the `reader.GetString()`, I'd be forced to add another line just to consume that position in the tempReader so I could read the following PropertyName, to check if it's $values.